### PR TITLE
RFP for 2019-Q4-PyPI

### PR DESCRIPTION
**Pull request by ewdurbin**

### CODE REVIEW

File: 2019-Q4-PyPI/RFP.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  - Referencing contributions to Free/Open Source projects is encouraged.
+
+- Project timeline estimates by milestone and task. These timelines should fit within our [project timeline](#project-timeline), however proposers may suggest adjustments based on their availability.
+  - Milestone 1 - Verifiable cryptographic signing of artifacts
+  - Milestone 2 - Systems for Automated Detection of Malicious Uploads
+
+- Project budget by milestone and task. Deviations from our [estimated budgets and caps](#estimated-budgets-and-caps) should be described and supported.
+  - Milestone 1 - Verifiable cryptographic signing of artifacts
+  - Milestone 2 - Systems for Automated Detection of Malicious Uploads
+
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: I suggest we give them an approximate page range. Such as:

"In total, we expect your proposal will be 5-10 pages long, but feel free to go under 5 or over 10 pages if you feel it's appropriate."

File: 2019-Q4-PyPI/RFP.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+## Evaluation criteria
+
+### Potential pass/fail:
+
+ - Contains all elements specified in the [elements of the proposal](#elements-of-proposal).
+ - Proposal is detailed enough to properly assess further criteria.
+ - Formatting and Submission requirements:
+   - Portable Document Format (PDF)
+   - Emailed to [ernest@python.org](mailto:ernest@python.org) by **October 16, 2019 [AoE](https://www.timeanddate.com/time/zones/aoe)** (2019-10-16T12:00:00Z)
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: Not your pyfound address? (OK if so! just checking.)

File: 2019-Q4-PyPI/RFP.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+* Development of systems for serving signed metadata compliant with [TUF](https://theupdateframework.github.io/)
+* Development of systems for appropriately caching and invalidation of caches for metadata service
+* Documentation for Administrators of PyPI for handling key material and interacting with the systems
+* Documentation for PyPI client libraries for sourcing and validating [TUF](https://theupdateframework.github.io/) metadata
+
+### Milestone 2 - Systems for Automated Detection of Malicious Uploads
+
+#### Summary
+
+* Development of infrastructure as either a standalone service or code in the [warehouse codebase](https://github.com/pypa/warehouse) to automatically screen metadata and uploads to PyPI for malicious content via pluggable checks, reporting results to PyPI
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: should be:

> Warehouse codebase

here and elsewhere in the document.

File: 2019-Q4-PyPI/RFP.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+Discussions leading to these milestones can be read in the [forum for our Request for Information](https://discuss.python.org/c/python-software-foundation/pypi-q4-rfi).
+
+### Milestone 1 - Verifiable cryptographic signing of artifacts
+
+#### Summary
+
+* Implementation of [PEP 458 once accepted](https://www.python.org/dev/peps/pep-0458/) to add integration of [The Update Framework](https://theupdateframework.github.io/) to PyPI
+* Development of either a stand alone service or code in the [warehouse codebase](https://github.com/pypa/warehouse) to create, sign, serve, and handle caching concerns for TUF metadata
+* Development of necessary code in the [warehouse codebase] to integrate TUF metadata and signing
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: Missing hyperlink here.

File: 2019-Q4-PyPI/RFP.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+- Do the proposers agree to the [project management and reporting](#project-management-and-reporting) requirements?
+- Are the project timeline estimates within our [project timeline](#project-timeline)? Are deviations from our timeline explained and supported?
+
+## Scope
+
+This Request for Proposals is seeking **backend developers** to implement, test, verify, and assist in the rollout of the following features to the codebase that powers [PyPI](https://pypi.org/).
+
+Discussions leading to these milestones can be read in the [forum for our Request for Information](https://discuss.python.org/c/python-software-foundation/pypi-q4-rfi).
+
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: I'd like to add a note here about what is NOT in scope:

"Not in scope: adding TUF metadata generation or reading functionality elsewhere in the packaging toolchain (setuptools, pip, etc.) or surfacing malware detection results to PyPI packagers and downloading users via the web UI or API"

File: 2019-Q4-PyPI/RFP.md (hunk 1 above)
Comment (ewdurbin) [reply 2019-Q4-PyPI/RFP.md]: I'll run through proposals from OTF RFP and get a feel for what solid proposals had.

File: 2019-Q4-PyPI/RFP.md (hunk 2 above)
Comment (ewdurbin) [reply 2019-Q4-PyPI/RFP.md]: That's actually a pretty good idea to better ensure deliverability. Thanks for the suggestion.

File: 2019-Q4-PyPI/RFP.md (hunk 1 above)
Comment (ewdurbin) [reply 2019-Q4-PyPI/RFP.md]: Your numbers were spot on :-D

**Comment by ewdurbin** [comment]
@brainwane ready for another review.

File: 2019-Q4-PyPI/RFP.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+|September 24|Request for Proposal period opened.|
+|October 16|Request for Proposal period closes.|
+|October 29|Date proposals will have received a decision.|
+|November 30|Contracts for accepted proposals should be finalized.|
+|December 2|Contract work commences.|
+
+
+## What is the Request for Proposals period?
+
+**A Request for Proposals (RFP)** is a process intended to allow us (the Python Software Foundation)to collect proposals from potential contractors and select contractor(s) best suited to fulfill the specified work.
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: need space after closing paren

File: 2019-Q4-PyPI/RFP.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+Discussions leading to these milestones can be read in the [forum for our Request for Information](https://discuss.python.org/c/python-software-foundation/pypi-q4-rfi).
+
+### Milestone 1 - Verifiable cryptographic signing of artifacts
+
+#### Summary
+
+* Implementation of [PEP 458 once accepted](https://www.python.org/dev/peps/pep-0458/) to add integration of [The Update Framework](https://theupdateframework.github.io/) to PyPI
+* Development of either a stand alone service or code in the [Warehouse codebase](https://github.com/pypa/warehouse) to create, sign, serve, and handle caching concerns for TUF metadata
+* Development of necessary code in the [Warehouse codebase] to integrate TUF](https://theupdateframework.github.io/) metadata and signing
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: Missing hyperlink for "Warehouse codebase" and missing opening bracket for TUF

File: 2019-Q4-PyPI/RFP.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+[Relevant background and context](https://github.com/pypa/warehouse/issues/5247)
+
+#### Specific tasks
+
+* Development of systems for generation and signing of metadata compliant with [TUF](https://theupdateframework.github.io/)
+* Development of systems for serving signed metadata compliant with [TUF](https://theupdateframework.github.io/)
+* Development of systems for appropriately caching and invalidation of caches for metadata service
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: How about:

> Development of systems for appropriately caching, and invalidating cache for, metadata service

File: 2019-Q4-PyPI/RFP.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+#### Summary
+
+* Development of infrastructure as either a standalone service or code in the [Warehouse codebase](https://github.com/pypa/warehouse) to automatically screen metadata and uploads to PyPI for malicious content via pluggable checks, reporting results to PyPI
+* Development of necessary code in the [Warehouse codebase](https://github.com/pypa/warehouse) to store results of automated screening for administrator review
+
+[Relevant background and context](https://github.com/pypa/warehouse/issues/5117)
+
+#### Specific tasks
+
+* Develop systems for running automated checks on new Projects, Release, and Release Files.
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: maybe should be

> Projects, Releases, and Release Files

File: 2019-Q4-PyPI/RFP.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+* Develop a pluggable system for development and deployment of new checks
+* Implement database models to store check results relating to Projects, Releases, and Release Files from checks
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: redundancy here? Perhaps:

> Implement database models to store check results relating to Projects, Releases, and Release Files

File: 2019-Q4-PyPI/RFP.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+* Implement Administrator views for reviewing results
+* Documentation process for the development and submission of new checks
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: A process, or just the documentation itself?

File: 2019-Q4-PyPI/RFP.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+**Note**: Surfacing malware detection results to PyPI publishers and users via the Web User Interface or any API is not in scope.
+
+## Estimated Budgets and Caps
+
+**Budgets and caps** are provided to help contractors in preparing their proposals.
+
+Budgets are PSF's estimates, and caps are based on the funding commitment that we have received.
+
+Caps are provided to indicate that we understand that some features may require more funds than our estimated budget. Proposals may go over budget up to the cap for one feature, but fall under budget for another.
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: I suggest we open up the possibility that a proposal actually goes way under the cap for 1 task and over the cap for the other task (as long as the proposal still includes a healthy documentation line item, and the total proposal is <= $65,000). Based on my conversations with Justin, I think it's a strong possibility that TUF implementation is actually going to be way faster than we've budgeted for.

File: 2019-Q4-PyPI/RFP.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+Please read this RFP and respond to let us know if you have questions, or submit a proposal if you are interested in performing the work.
+
+## Timeline
+
+|Date|Milestone|
+|---|---|
+|August 28|Request for Information period opened.|
+|September 18|Request for Information period closed.|
+|September 24|Request for Proposal period opened.|
```
Comment (brainwane) [inline 2019-Q4-PyPI/RFP.md]: Note to update this to the current date when merging and publishing.

File: 2019-Q4-PyPI/RFP.md (hunk 12 above)
Comment (ewdurbin) [reply 2019-Q4-PyPI/RFP.md]: 👍

**Comment by ewdurbin** [comment]
The only concern raised is that we need to get PEP 458 revised as necessary to pin down specifics such as key distribution (who, where, how many?) and any technical choices that TUF leaves up to implementations. We'll need to work to revise 458 and have it accepted by inviting experts and those who have previously implemented TUF.

**Pull request merged**